### PR TITLE
Pull Request: Up-And-Running Supervised Training Example

### PR DESCRIPTION
**Pull request by EtienneChollet**
## Overview
This PR introduces enhancements and new features, focusing on I/O modules, optimizers, loss functions, datasets, and documentation - and generally, getting everything up and running for some basic supervised training!

## Examples
Two examples of supervised training have been included in `cassetta/docs/examples`:
1. `supervised-segmentation.ipynb`: Demonstrating the initial process of defining and training a model from scratch, and saving it's checkpoints to an experiment directory.
2. `supervised-finetuning.ipynb`: Demonstrating how one might continue training (fine-tuning) a model from saved checkpoints.

### CODE REVIEW

**Comment by balbasty** [comment]
@EtienneChollet tests are failing because you're using the yaml package but it's not a dependency. 

I am not entirely usre whether we want it as a dependency. Can you maybe only import it inside the function where it's use (which I guess is only when the filename ends with ".yaml") and throw a nice Error if the package is not available?
